### PR TITLE
Add signal strengths handling to assignments to nets of size 1

### PR DESCRIPTION
**Pull request by RRozak**
I add support for multiple continuous assignments (and logic gates parsed as assignments, for example `or`) with size of 1.
Since signal strength is rarely used, I wanted to use approach that don't increase memory cost of simulation when they aren't used.
This PR needs (and contains) changes from https://github.com/verilator/verilator/pull/3545.
I created separate visitor to handle signal strengths, but this handling can be moved to some other visitor that already exists.

Description of my approach (it is also in V3SignalStrength.cpp):
The approach is to create var__s0 and var__s1 for each var that is on LHS of at least two
assignments or its assignment has strength highz0 or highz1, which may change its value. var__s0
is used for storing strength of the strongest assignment of value 0 to var, var__s1 stores
strength of the strongest assigment of value 1. We want to replace the current assignments to
the var with one assignment with the RHS depending on __s0 and __s1.

First we iterate through the list of assignments that have the var on the LHS.
We want to do the following steps, but we may not know the value of the RHS at verilation time,
so we need to do all of them and add a comparison of the RHS with the value corresponding to the
step.
* If the RHS has value 0, we try to update __s0
* If the RHS has value 1, we try to update __s1
* If the RHS has value x, we try to update both __s vars (to handle the case when
assignment of x is the strongest)

If the RHS has value z, no __s vars are touched (it is the weakest signal and any other value will
overwrite it), so we can skip this case.

After the iteration we create new assignment to var, according to values of __s0 and __s1:
s0 > s1:
var = 0
s0 < s1:
var = 1
s0 == s1 and s0 != 0:
var = x
s0 == s1 and s0 == 0:
var = z

### CODE REVIEW

File: src/V3SignalStrength.cpp
Comment (kbieganski) [inline src/V3SignalStrength.cpp:64]: Do we need to store this in a map? I think this can be done in `visit(AstAssignW*)`, while iterating the assigns. Utilize the `user{1,2,3,4,5}` pointers.

File: src/V3SignalStrength.cpp
Comment (kbieganski) [inline src/V3SignalStrength.cpp:106]: So do we only handle 1-bit signals? Or am I missing something?

**Review by kbieganski — CHANGES_REQUESTED** [review]

File: src/V3AstNodes.cpp
Comment (RRozak) [reply src/V3AstNodes.cpp:1286]: This is unused. I just forgot to remove it.

File: src/V3ParseGrammar.cpp
Comment (RRozak) [reply src/V3ParseGrammar.cpp:86]: This value can only be 0 or 1. And strength can only be supply0 or supply1, respectively.

File: src/V3SignalStrength.cpp
Comment (RRozak) [reply src/V3SignalStrength.cpp:104]: Yes, for now only 1-bit signals are supported. I am working on multidimensional now.

File: src/V3SignalStrength.cpp
Comment (RRozak) [reply src/V3SignalStrength.cpp:64]: But there can be more than 5 nets in one module.

File: src/V3SignalStrength.cpp
Comment (RRozak) [reply src/V3SignalStrength.cpp:31]: I thought that I will add support for other nodes. I created separate visitor to be more flexible. Now I know that I will need to handle for now assignments only. I will move it to other visitor once signal strength on multidimensional nets is supported.

File: src/V3ParseImp.cpp
Comment (RRozak) [reply src/V3ParseImp.cpp:402]: Done

File: src/V3AstNodes.cpp
Comment (kbieganski) [inline src/V3AstNodes.cpp:1290]:
```suggestion
    char valChar = val ? '1' : '0';
    switch (m_level) {
    case HIGHZ: return "highz" + valChar;
    case SMALL: return "small" + valChar;
```
and so on.

Though the Verilator maintainers may want this to look like other enums, e.g. `VEdgeType`. See the `ascii` method there.

File: src/V3AstNodes.h
Comment (kbieganski) [inline src/V3AstNodes.h:297]:
```suggestion
    bool m_val;
```

File: src/V3AstNodes.cpp
Comment (kbieganski) [inline src/V3AstNodes.cpp:1302]: AST constructors are all in the header file.

File: src/V3ParseGrammar.cpp
Comment (kbieganski) [inline src/V3ParseGrammar.cpp:92]:
```suggestion
        = new AstAssignW{fileline, new AstVarRef{fileline, name, VAccess::WRITE},
                         new AstConst{fileline, AstConst::StringToParse{}, (value ? "'1" : "'0")}};
    AstStrengthSpec* strengthSpecp
        = new AstStrengthSpec{fileline, new AstStrength{fileline, AstStrength::SUPPLY, 0},
                              new AstStrength{fileline, AstStrength::SUPPLY, 1}};
```
This is the new Verilator style. They're going to request that, so you may as well do it now. Do it in the rest of the patch as well.

File: src/V3SignalStrength.cpp
Comment (kbieganski) [reply src/V3SignalStrength.cpp:116]: That's why I said this can be done in `visit(AstAssignW*)` (that inner loop at least), not in `visit(AstNodeModule*)`. The new vars and the conditional `AssignW` can be created in `V3SignalStrength.cpp:76`. Put the shared info in `user*`. Let's not re-iterate over nodes if we can avoid it. See how other visitors utilize this, e.g. `V3Delayed` puts objects there for future re-use.

I may be wrong though, and this is needed.

File: src/V3SignalStrength.cpp
Comment (kbieganski) [inline src/V3SignalStrength.cpp:77]:
```suggestion
            if (isWire && !isRanged) {
                m_assigns[varRefp->varp()].push_back(nodep);
            } else if (nodep->strengthSpecp()) {
```

File: src/V3SignalStrength.cpp
Comment (kbieganski) [inline src/V3SignalStrength.cpp:117]: If you keep this, make this a reference.

File: src/V3SignalStrength.cpp
Comment (kbieganski) [inline src/V3SignalStrength.cpp:122]:
```suggestion
                AstVar* const varp = varpAssigns.first;
```
Likewise for other such declarations.

File: src/V3SignalStrength.h
Comment (kbieganski) [inline src/V3SignalStrength.h:29]:
```suggestion
    static void strengthAll(AstNetlist* nodep);
```
Consistent with other such methods (`constifyAll`, `assertAll`, etc)

File: src/V3Tristate.cpp
Comment (kbieganski) [inline src/V3Tristate.cpp:415]:
```suggestion
    AstConst* getEnConstp(AstConst* constp) {
```
I think the `p` at the end is enough? Even that might not be necessary.

File: src/V3SignalStrength.cpp
Comment (kbieganski) [inline src/V3SignalStrength.cpp:200]: If you keep this, split it into methods so it's easier to read.

File: src/verilog.y
Comment (kbieganski) [inline src/verilog.y:5062]: Formatting, but I guess you're working on it?

**Review by kbieganski — CHANGES_REQUESTED** [review]

File: src/verilog.y
Comment (RRozak) [reply src/verilog.y:5062]: I knot that formatting needs to be changed, but I don't know which tool to use and what the configuration should be. Maybe we can just ask Wilson about it. I decided to leave it for now.

File: src/V3AstNodes.h
Comment (RRozak) [reply src/V3AstNodes.h]: Done

File: src/V3AstNodes.cpp
Comment (RRozak) [reply src/V3AstNodes.cpp]: Done

File: src/V3SignalStrength.h
Comment (RRozak) [reply src/V3SignalStrength.h]: Done

File: src/V3Tristate.cpp
Comment (RRozak) [reply src/V3Tristate.cpp]: Done. I updated https://github.com/verilator/verilator/pull/3545.

File: src/V3SignalStrength.cpp
Comment (RRozak) [reply src/V3SignalStrength.cpp:116]: I know what you mean now. This would require to store `__s0`, `__s1` variables, their scope (or we can use just one for computing strengths for all of variables) and first assignment to that variable (to remove it when we encounter second assignment), so we would need to use 3 `user*p` pointers.

We know that these `__s*` variables will be needed at the same moment when we should remove the first assignment, so it is possible to use the same `user*p` pointer for assignment and for one of the variables, but it will make the situation that meaning of that pointer will change during handling of 2nd assignment. So let's say that 2 using of 2 userp pointers is the minimum.

As we previously talked, it may be required to move content of this visitor to some already existing visitor. Using of `user*p` pointers will make it more complicated. On the other hand, as I checked, V3Width is using none of these user*p, so maybe we don't need to care about it, if I have to move this to exactly V3Width visitor.

In `visit(AstNodeModule*)` we handle for 2nd time only these assignment that require any work (they have signal strengths or are one of multiple assignments to the same variable). I don't know how often it happens, but handling of them involve creating additional expressions, which is also a cost.

So I think it is possible to use `user*p`, but I think that benefits from it will be small and it will complicate the code. What do you think?

File: src/V3SignalStrength.cpp
Comment (RRozak) [reply src/V3SignalStrength.cpp]: Done.

File: src/V3SignalStrength.cpp
Comment (RRozak) [reply src/V3SignalStrength.cpp]: Done

File: src/V3SignalStrength.cpp
Comment (RRozak) [reply src/V3SignalStrength.cpp]: Done

File: src/V3SignalStrength.cpp
Comment (RRozak) [reply src/V3SignalStrength.cpp]: Done

File: src/V3AstNodes.cpp
Comment (RRozak) [reply src/V3AstNodes.cpp]: Done. I also get rid of `val` field.

File: src/V3SignalStrength.cpp
Comment (kbieganski) [reply src/V3SignalStrength.cpp:192]: Ok, let's leave it up to Verilator maintainers.

**Review by kbieganski — CHANGES_REQUESTED** [review]

File: src/V3Tristate.cpp
Comment (RRozak) [reply src/V3Tristate.cpp:970]: Here it can't be `const`, because we assign a proper node then.

File: src/V3SignalStrength.cpp
Comment (RRozak) [reply src/V3SignalStrength.cpp]: Done

File: src/V3Tristate.cpp
Comment (RRozak) [reply src/V3Tristate.cpp]: Done

**Comment by RRozak** [comment]
According to Wilson's suggestion, I changed the approach and continue work here: https://github.com/antmicro/verilator-1/pull/18.

**Pull request closed without merging**
